### DISCUSSION TITLE
Welcome to underscore-java Discussions!

### DISCUSSION THREAD

**Discussion by javadev** [javadev/underscore-java#277] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by shreyamns** [comment]
_**What if we have to convert JSON into attribute-based XML. I have tried using the jsonToXml function but it is giving as element-based XML.**_

Example->
**JSON->**
```
{"author": { 
    "extn": {"externalSystemCode": ""}, 
    "EMailID": "", 
     "DOB": "", 
     "PlaceId": "", 
     "SSN": "" 
}}
```

**Expected XML->**

```
<?xml version="1.0" encoding="UTF-8"?>
<RootElm>
    <author DOB="" EMailID="" PlaceId="" SSN="">
        <extn externalSystemCode=""/>
    </author>
</RootElm>
```

@javadev could you please help.

**Reply by javadev** [reply]
You may add minus to keys. It will generate attributes.

**Reply by javadev** [reply]
You may use this json to get the required XML.
```json
{
  "RootElm": {
    "author": {
      "-DOB": "",
      "-EMailID": "",
      "-PlaceId": "",
      "-SSN": "",
      "extn": {
        "-externalSystemCode": "",
        "-self-closing": "true"
      }
    }
  }
}
```

**Reply by shreyamns** [reply]
Thanks for replying @javadev. Apart from using a hyphen, is there any other way to convert JSON into attribute-based XML? Because I am getting JSON object which is without the hyphen. And converting every attribute with the hyphen would be a tedious task. So if you can suggest some other method that would be great. Or is it possible that using underscore library? I can convert our JSON into this form.

**Reply by javadev** [reply]
We may add a new mode to the jsonToXml method FORCE_ATTRIBUTE_USAGE.

**Reply by shreyamns** [reply]
Hi @javadev, So how can we use FORCE_ATTRIBUTE_USAGE. Is there any function already available that will handle the conversion?

**Reply by javadev** [reply]
We have several modes for xmlToJson method. I suggest to add new for jsonToXml.

**Reply by javadev** [reply]
I added a new mode. https://github.com/javadev/underscore-java/pull/279

**Reply by shreyamns** [reply]
Thank you so much. It is really helpful for me, but the recent changes are not getting reflected while adding the dependency. The new method for mode is not coming.

**Reply by javadev** [reply]
The new version was released today. 🎉

**Reply by shreyamns** [reply]
Thank you so much for the help @javadev. Now I am able to use the new method. Thanks a ton. 👍

**Reply by shreyamns** [reply]
Hi @javadev ,
    I am curious to know that is there any method which is available in the underscore library which will change the root tag name. As of now, every XML converted with root tag name as root. So if we want to customize it can we do this?

**Reply by javadev** [reply]
Hi @shreyamns. We have a rename() method. It may be a new mode to convert json to XML with a new root tag.

**Reply by javadev** [reply]
I added a new method U.jsonToXml(json, newRootName). https://github.com/javadev/underscore-java/pull/281

**Reply by shreyamns** [reply]
Thank you so much @javadev. I have one question. if I need to use both the mode simultaneously how can I use it? Because in code I can see at a time only a single mode can be used. If you can have one more if-else statement for force attribute usage along with a defined root name that would be grateful.

**Reply by javadev** [reply]
I added a new mode FORCE_ATTRIBUTE_USAGE_AND_DEFINE_ROOT_NAME. https://github.com/javadev/underscore-java/pull/282

**Reply by shreyamns** [reply]
Thanks much @javadev  when will the next release be available? 

Also, is there any method available such a way that we can add some custom attribute to the XML after JSON to attribute-based XML conversion? 

For example: 
`<author DOB="" EMailID="" PlaceId="" SSN=""> <extn externalSystemCode=""> </author>`

I want to add one attribute(name) at author level to have the output as below: 

`<author name="" DOB="" EMailID="" PlaceId="" SSN=""> <extn externalSystemCode=""> </author>`

**Reply by javadev** [reply]
I tested this code.

```java
U.objectBuilder()
            .fromXml("<author DOB=\"\" EMailID=\"\" PlaceId=\"\" SSN=\"\"> <extn externalSystemCode=\"\"/> </author>")
            .set("author.-name", "").toXml()
```
the resulted XML:
```xml
<author DOB="" EMailID="" PlaceId="" SSN="" name="">
  <extn externalSystemCode=""/>
</author>
```

**Reply by javadev** [reply]
When will the next release be available?

The new version will be released in March.

**Reply by shreyamns** [reply]
Thanks for the help @javadev. Will be waiting for the next release.

**Reply by javadev** [reply]
The new version was released today. 🎉

**Comment by moussabenbouzid** [comment]
Hello,
Thank you very much for this amazing library.
I still have one request and it’s about a json conversion  : 
I have to convert json maps to xml, and the keys of the map start with a number...
Example :
{"1": {
      "field" : "value"
       },
"2": {
"field" : "value2
       }
Sometimes, i have also keys with spaces ...
But an XML tag cannot neither start with a number nor contain spaces.
The U.jsonToXml method handles the issue by replacing the tags starting with a number with something like "__HA__" or "__GE__" and spaces with "__EA__".
I don't really understand the reasoning. Is there a documention to explain how it works ?
Best regards

**Reply by javadev** [reply]
Hi @moussabenbouzid 

Thank you for the question.

We use base32 algorithm to support numbers and special characters.

**Comment by moussabenbouzid** [comment]
Hello, 
Thank you very much !

I have another question : 
When converting a json map with a single element, I discovered that there is no root element. Is there a configuration to have it in all cases?

Best regards

**Reply by javadev** [reply]
We have a mode to add root element while converting U.JsonToXmlMode.ADD_ROOT.

**Comment by moussabenbouzid** [comment]
Hello, 
Thank you !

That works well 👍  
Is it possible to combine several modes in the same transformation? For example, the forcing root mode and the replace null with empty values.

Best regards

**Reply by javadev** [reply]
You may convert json to map, execute trasformation methods and then convert map to xml.
